doc/wdel.1
.TH "wdel" "1" "0.6" "Hagen Fritsch, Jan Larres" "Internet Applications - FTP"
.SH "NAME"
wdel \- A wget\-like program for deleting files on ftp servers
.SH "SYNOPSIS"
wdel [\fIoption\fR]... [\fI\s-1URL\s0\fR [\fI\s-1file\s0\fR]...]...
.SH "DESCRIPTION"
Wdel is a free utility that is able to delete files on an ftp server.
.PP
Wdel is non\-interactive and background-capable. It can delete files or whole
directories and is meant to be a robust client even for unstable connections
and will therefore retry to delete a file, if the connection broke.
.PP
Wdel supports connections through proxies, allowing you to use it in an
environment that can access the internet only via a proxy or to provide
anonymity by hiding your ip\-address to the server.
For SOCKSv5\-proxies Wdel supports also listening mode, allowing you to use
port-mode ftp through a proxy (useful if the remote ftp is behind a firewall
or a gateway).
.SH "URL\-Input\-Handling"
URLs are recognized by the ftp://\-prefix.
.PP
Wdel first reads the URLs from the command line, and associates every file
with the last URL until a new URL is found (note that this is different
from wput). Then, it reads the \-\-input\-file (if any) and parses the URLs
in the same way.  So you can specify e.g. one URL and read all filenames
from a file.
.PP
If there are no explicit filenames given for a singe URL, then the last
part of the URL is taken to be the file/directory to be deleted. If there
are filenames, then the last part of the URL is considered to be the
directory where these files reside in. Slashes at the end of the URL or the
filename (in case it is a directory to be deleted) are ignored.
.SH "OPTIONS"
.Sh "Basic Startup Options"
.IP "\fB\-V\fR" 4
.IX Item "-V"
.PD 0
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
.PD
Display the version of wdel.
.IP "\fB\-h\fR" 4
.IX Item "-h"
.PD 0
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
.PD
Print a help screen, with a short description of wdel's command-line options.
.IP "\fB\-b\fR" 4
.IX Item "-b"
.PD 0
.IP "\fB\-\-background\fR" 4
.IX Item "--background"
.PD
Go to background immediately after startup.  If no output file is given,
wdel will redirect its output to "./wdel-log"
.Sh "Logging and Input File Options"
.IX Subsection "Logging and Input File Options"
.IP "\fB\-o\fR \fIlogfile\fR" 4
.IX Item "-o logfile"
.PD 0
.IP "\fB\-\-output\-file=\fR\fIlogfile\fR" 4
.IX Item "--output-file=logfile"
.PD
Log all messages to \fIlogfile\fR.
.IP "\fB\-a\fR \fIlogfile\fR" 4
.IX Item "-a logfile"
.PD 0
.IP "\fB\-\-append\-output=\fR\fIlogfile\fR" 4
.IX Item "--append-output=logfile"
.PD
Append all logged messages to \fIlogfile\fR.
.IP "\fB\-q\fR" 4
.IX Item "-q"
.PD 0
.IP "\fB\-\-quiet\fR" 4
.IX Item "--quiet"
.PD
Turn off Wdel's output.
.IP "\fB\-v\fR" 4
.IX Item "-v"
.PD 0
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
.PD
Turn on verbose output. This gives some more information about what Wdel
does. If you specify this flag twice, you get debug output.
.IP "\fB\-nv\fR" 4
.IX Item "-nv"
.PD 0
.IP "\fB\-\-less\-verbose\fR" 4
.IX Item "--less-verbose"
.PD
Be less verbose. That means reducing Wdel's output to a minimun. Specifying
this flag more often is equal to the --quiet flag.
.IP "\fB\-i\fR \fIfile\fR" 4
.IX Item "-i file"
.PD 0
.IP "\fB\-\-input\-file=\fR\fIfile\fR" 4
.IX Item "--input-file=file"
.PD
Reads URLs and filenames from \fIfile\fR. These will be added to the queue
just like if they were supplied via the command line.  See also the
URL-Input-Handling section.
.PD
If \fIfile\fR is \-, the URLs will be read from stdin.
.IP "\fB\-s\fR" 4
.IX Item "-s"
.PD 0
.IP "\fB\-\-sort\fR" 4
.IX Item "--sort"
.PD
If sorting is enabled Wdel first reads all URLs from any input-devices available
and will sort them before deleting each file.
.PD
The sorting order is: ip/hostname, port, username, password, directory, filename.
Sorting requires a bit more memory since all data needs to be held there.
.IP "\fB\-\-basename=\fR\fIpath\fR" 4
.IX Item "--basename=path"
.PD
This option causes Wdel to snip \fIpath\fR from all filenames when they are
connected to the URL. wdel ftp://host/ /usr/share/doc.tgz would delete
ftp://host//usr/share/doc.tgz, whereas specifing /usr/share/ as basename will
result in ftp://host/doc.tgz being deleted.
.Sh "Connection Options"
.IX Subsection "Connection Options"
.IP "\fB\-\-bind\-address=\fR\fI\s-1ADDRESS\s0\fR" 4
.IX Item "--bind-address=ADDRESS"
When making client \s-1TCP/IP\s0 connections, \f(CW\*(C`bind()\*(C'\fR to \fI\s-1ADDRESS\s0\fR on
the local machine.  \fI\s-1ADDRESS\s0\fR may be specified as a hostname or \s-1IP\s0
address.  This option can be useful if your machine is bound to multiple
IPs.
.IP "\fB\-t\fR \fInumber\fR" 4
.IX Item "-t number"
.PD 0
.IP "\fB\-\-tries=\fR\fInumber\fR" 4
.IX Item "--tries=number"
.PD
Set number of retries to \fInumber\fR.  Specify -1 for infinite retrying.
The default is to retry 3 times, with some exceptions on
fatal errors.
.IP "\fB\-Y\fR \fIMODE\fR" 4
.IX Item "-Y MODE"
.PD 0
.IP "\fB\-\-proxy=\fR\fIMODE\fR" 4
.IX Item "--proxy=MODE"
.PD
MODE can be either \fIhttp\fR for http-based proxies (such as SQUID),
\fIsocks\fR for SOCKSv5 proxies or \fIoff\fR to disable the proxy.
.IP "\fB\-\-proxy-user=\fR\fINAME\fR" 4
.IX Item "--proxy-user=NAME"
If the proxy\-server requires authentication, use \fINAME\fR as user-name.
You need to specify \-\-proxy-pass too. These information can also be
stored in the wputrc\-file.
.IP "\fB\-\-proxy-pass=\fR\fIPASS\fR" 4
.IX Item "--proxy-pass=PASS"
Specifies the password to use for the proxy.
.Sh "FTP Options"
.IX Subsection "FTP Options"
.IP "\fB\-p\fR" 4
.IX Item "-p"
.PD 0
.IP "\fB\-\-port\-mode\fR" 4
.IX Item "--port-mode"
Per default, Wdel uses passive mode ftp, which works well for most
configurations. If passive mode fails, Wdel automatically falls back to
port mode.
.PD
If you want Wdel to start using port mode ftp, specify this flag.
.IP "\fB\-\-force-tls\fR" 4
.IX Item "--force-tls"
If this flag is specified and Wdel is linked with the OpenSSL-library, the flag

ChangeLog
Sat Jul 26 18:33:32 CEST 2008 Jan Larres
	* add a wdel program for deleting files/directories on a server and
	  a corresponding man page

Fri Jul 25 22:32:49 CEST 2008 Jan Larres
	* add support for the SITE CHMOD command

Wed Jul 02 13:19:21 CET 2008
	* fixing #2007526 (now using EINPROGRESS and not 115)

Wed Dec 11 20:53:52 CET 2007 wput-0.6.1
	* adding russian (ru) translation by Vyacheslav Nikitin
	* building now to wput-VERSION fixing #1844014

Sun Dec 02 16:55:29 CET 2007
	* fixing #1776300 (build issues), #1834977 (netrc),
	  #1737168 (return values), #1651042 (resume-corruption),
	  #1599242 (TYPE<->SIZE), #1525208 (ps and passwords),
	  #1523072 (timeoffset)
	* adding simplified chinese (zh) translation by Sender Soff
	* adding turkish (tr) translation by Eren Türkay

Wed Apr 12 18:17:32 CET 2006
	* adding dutch (nl) translation by Sebastiaan van der Velden

Sat Apr 08 15:11:50 CET 2006
	* gnutls created SIGPIPE events when the connection broke
	  down. ignoring SIGPIPE now
	* fixing some issues in transferrate/eta-calculation for
	  slower rates than 10K

Fri Apr 07 14:12:33 CET 2006
	* fixed the bar_update function which was too busy printing
	  out updates. limited to 10 updates per second
	* fixed the bar_creating function which called get_terminal_size
	  even in background mode, causing strange errors afterwards

Thu Apr 06 23:05:49 CET 2006
	* changed SOCKET_ERROR to SOCK_ERROR because SOCKET_ERROR
	  is somewhat defined by winsock in w32
	* minor other changes to help compilation for win32

Wed Apr 05 12:57:10 CET 2006
	* fixing a bug in get_relative_path which returned
	  false data for pathcombinations like "a","abc"

Tue Apr 04 18:46:34 CET 2006 wput-0.6
	* using gnutls now. adjusting includes. no need for win32
	* parsing of mdtm forgot to decrease the month. fixed
	* setting target_fsize to -1 to indicate unexisting files,
	  so that wput will upload even files with 0 filesize
	* fixing double free() of hostname if the host could not be
	  resolved
	* fixing reported bug in Makefile
	* adding support for .netrc-files as suggested by a patch
	  of Alexander Pohoyda
	* when sending or parsing PORT/PASV commands, now using
	  htons to convert to network byte order

Thu May 26 09:30:54 CET 2005
	* buffer-allocation in http-proxy-module forgot to count five
	  additional bytes for the port-value which led to segfaults

Sun May  1 10:00:16 CET 2005
	* fixed a calculation for calculate_transfer_rate that caused an
	  data-overflow for rates beyond 210 KiB/s
	* another fix for set_resuming because the last fix made things even
	  worse. works fine now ;)

Fri Apr 29 16:53:14 CET 2005
	* minor fix in set_resuming which mixed up some cases

Wed Mar 23 10:32:41 CET 2005
	* fix for base64-routine which forgot to set the last character to \0

Sat Feb  5 20:37:09 CET 2005 wput-pre0.6
	* making compilation for windows work (ssllib in dll, some
	  modifications (sslib_in_use flag), strptime included, some
	  fixes for *nix-header-files)

Tue Feb  2 22:10:45 CET 2005

src/wdel.c
/* Author: Hagen Fritsch <fritsch+wput-src@in.tum.de>,
           Jan Larres <jan@majutsushi.net>
   (C) 2002-2008 by Hagen Fritsch, Jan Larres

   This file is part of wput.

   This programm is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License 
   as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   The wput is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

   You should have received a copy of the GNU General Public
   License along with the wput; if not, write to the Free
   Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
   02111-1307 USA.  */

/* command-line-parsing routines and core control functions */

#include <sys/types.h>
#include <sys/stat.h>
#ifndef WIN32
#  include <netinet/in.h>
#endif

#include "wput.h"
#include "netrc.h"

#ifdef WIN32
#  include "getopt/getopt.h"
#else
#  define _GNU_SOURCE
#  ifdef HAVE_GETOPT_H
#    include <getopt.h>
#  else
#    ifdef HAVE_GNUGETOPT_GETOPT_H
#      include <gnugetopt/getopt.h>
#    else
#      include "getopt/getopt.h"
#    endif
#  endif
#endif

#ifdef ENABLE_NLS
#  ifdef HAVE_LOCALE_H
#    include <locale.h>
#  endif
#endif

#include "progress.h"
#include "_queue.h"
#include "utils.h"

extern char *optarg;

#ifdef WIN32
const static char * version = "0.6-w32";
#else
const static char * version = "0.6";
#endif

_fsession * fsession_queue_entry_point = NULL;

void commandlineoptions(int argc, char * argv[]);
int  start_fsession();
int  start_ftp();
int  start_recur_ftp();
void read_netrc_file(void);

int main(int argc, char *argv[]){
#ifdef WIN32
    /* i don't know why, but if i call WSAStartup just once, there is an
     * error when initialising sockets later (10093). quite strange and
     * took me quite a while to figure out *dang* */
    WSADATA sa;
    if(WSAStartup(MAKEWORD(2,2),&sa) != 0) {
        fprintf(stderr, "1st: Error Initializing Windows Socket DLL (WSAStartup Failed).\nError-Code: 0x%x", GetLastError());
        exit(4);
    }
    if(WSAStartup(MAKEWORD(2,2),&sa) != 0) {
        fprintf(stderr, "2nd: Error Initializing Windows Socket DLL (WSAStartup Failed).\nError-Code: 0x%x", GetLastError());
        exit(4);
    }
#endif
    signal(SIGPIPE, SIG_IGN);
    /* i18n */
#ifdef ENABLE_NLS
    /* LC_MESSAGES is enough as the only thing that is done is
     * the translation of text-messages */
#  ifdef LC_MESSAGES
    setlocale (LC_MESSAGES, "");
    setlocale (LC_CTYPE, "");
#  else
    setlocale (LC_ALL, "");
#  endif
    /* Set the text message domain.  */
    bindtextdomain ("wput", LOCALEDIR);
    textdomain ("wput");
#endif


    /* initialise global options and set default values */
    memset(&opt, 0, sizeof(opt));

    /* create the ssl framework */
#ifdef HAVE_SSL
    SSL_library_init();
#endif

    opt.sbuf      = malloc(82);
    opt.retry     = -1;
    opt.retry_interval = 10;
    opt.time_deviation = 10;
    opt.output    = stdout;

    opt.verbose   = vNORMAL;
    opt.bindaddr  = INADDR_ANY;
    opt.ps.bind   = 1;
    opt.session_start = wtimer_alloc();

    opt.email_address = cpy("wput@localhost.com");

    opt.wdel = 1;
    opt.no_directories = 1;

    /* env overrides home overrides system wputrc */
#ifdef SYSTEM_WPUTRC
    readwputrc(SYSTEM_WPUTRC);
#endif
    readwputrc(NULL); /*homedir*/
    readwputrc(getenv("WPUTRC"));

    parse_proxy(getenv("ftp_proxy"));

    read_netrc_file();
    read_password_file(getenv("PASSWORDFILE"));

    wtimer_reset(opt.session_start);

    commandlineoptions(argc, argv);

#ifndef WIN32
    /* If we are still at stdout, then redirect output to 'wdel-log'. */
    if(opt.background) {
        if(opt.output == stdout) {
                printout(vLESS, _("Resuming in background. Logging output to 'wdel-log'.\n"));
                opt.output   = fopen("wdel-log", "a");
                if(opt.output == NULL) { perror(_("Unable to open logfile")); exit(4); }
        } else
                puts("Resuming in background.\n");

        if(fork() > 0) exit(0);

        if(opt.input_pipe) {
            printout(vNORMAL, _("Warning: "));
            printout(vNORMAL, _("background-mode might not work correctly, if the input-pipe needs to read from stdin (like cat -).\n"));
        }
        if(opt.input != stdin) {
            /* create a new session and release the console-descriptors... */
            setsid ();
            freopen ("/dev/null", "r", stdin);
            freopen ("/dev/null", "w", stdout);
            freopen ("/dev/null", "w", stderr);
        } else {
            printout(vNORMAL, _("Warning: "));
            printout(vNORMAL, _("reading urls from stdin, while running in background-mode is not reliable.\n"));
        }
    }
#endif
    if((opt.ps.ip == 0 || opt.ps.port == 0) && opt.ps.type != PROXY_OFF) {
        printout(vNORMAL, _("Warning: "));
        printout(vNORMAL, _("Ignoring request to turn proxy-usage on, since no proxy is configured.\n"));
        opt.ps.type = PROXY_OFF;
    }

    /* read URLs from input-file */
    if(opt.input != 0) read_urls();

    /* separate the urls from a potential file and ensure that the urls end in a '/' */
    separate_urls();

    /* process all url/file combinations that were supplied by commandline */
    queue_process(0);

    /* now we've everything we need or are already done */
    if(opt.sorturls) {
        printout(vDEBUG, "Transmitting sorted fsessions\n");
        while(fsession_queue_entry_point != NULL) {
                int res = fsession_process_file(fsession_queue_entry_point, opt.curftp);
                if(res == -1)      opt.failed++;
                else if(res == -2) opt.skipped++;
                opt.curftp = fsession_queue_entry_point->ftp;
                free_fsession(fsession_queue_entry_point);
        }
    }

    /* finally close any existing connections */
    if(opt.curftp) ftp_quit(opt.curftp);

    if(opt.transfered == 0 && opt.skipped == 0 && opt.failed == 0)
        printout(vNORMAL, _("Nothing done. Try `wdel --help'.\n"));
    else
        printout(vNORMAL, _("FINISHED --%s--\n"), time_str());

    if(opt.transfered > 0)
        printout(vNORMAL, opt.transfered == 1 ?
        _("Deleted %d file\n") : 
        _("Deleted %d files\n"), 
            opt.transfered);
    if(opt.skipped > 0)
        printout(vNORMAL, opt.skipped == 1 ? _("Skipped %d file.\n") : _("Skipped %d files.\n"), opt.skipped);
    if(opt.failed > 0)
        printout(vNORMAL, opt.failed == 1 ? _("Deletion of %d file failed.\n") : _("Deletion of %d files failed.\n"), opt.failed);

    /* clean up */
    free(opt.session_start);
    free(opt.email_address);
    free(opt.sbuf);

    if(opt.ps.pass)    free(opt.ps.pass);
    if(opt.ps.user)    free(opt.ps.user);
    if(opt.last_url)   free(opt.last_url);

    if(opt.pl)         password_list_free(opt.pl);
    skiplist_free(opt.skipdlist);

#ifdef MEMDBG
    print_unfree();
#endif

    return ((opt.failed != 0) * 2) | (opt.skipped != 0);
}
/* read urls/files from input-file as though they were supplied as command-line-arguments */
int read_urls(void) {
    char * url;
    char * p;
    while ((url = p = read_line(opt.input))) {
        /* skip spaces and skip lines beginning with a # */
        printout(vDEBUG, "read `%s'\n", p);

        while( isspace(*p) ) p++;
        if(*p == '#') {
            free(url);
            continue;
        }

        /* just foolproof check for \r\n */
        if(url[strlen(url)-2] == '\r') 
            url[strlen(url)-2] = 0;
        url[strlen(url)-1] = 0;

        if(!strncmp(p, "ftp://", 6))
            wdel_queue_add_entry(NULL, cpy(p));
        else
            wdel_queue_add_file(cpy(p));

        free(url);
    }
    return 0;
}
/* ugly code to parse through the wputrc-options */
int set_option(char * com, char * val) {
  printout(vDEBUG, "Setting option '%s' to '%s'\n", com, val);

  switch(*com) {
  case '1':
  case '2':
      return 0;
  case 'b':
      if(!strncasecmp(com, "bind-address", 13)) {
            if(get_ip_addr(optarg, &opt.bindaddr) == -1) {
                printout(vMORE, _("Error: "));
                printout(vMORE, _("`%s' could not be resolved. "), optarg);
                printout(vLESS, _("Exiting.\n"));
                exit(4);
            }
            return 0;
        } else return -1;
  case 'c':
      if(!strncasecmp(com, "connection_mode", 16)) {
        if(!strncasecmp(val, "pasv", 5)) {
          opt.portmode = 0;
        } else if(!strncasecmp(val, "port", 4)) {
          opt.portmode = 1;
        } else return -1;
      }
      else if(!strncasecmp(com, "chmod", 6)) {
          return 0;
      } else return -1;
      return 0;
#ifdef HAVE_SSL
  case 'f':
      if(!strncmp(com, "force_tls", 9))
          opt.tls = !strncasecmp(val, "on", 3);
      else
          return -1;
      return 0;
#endif
  case 'm':
      if(!strncmp(com, "email_address", 13))
          opt.email_address = cpy(val);
      else return -1;
      return 0;
  case 'p':
      if(!strncasecmp(com, "proxy", 6)) {
        if(!strncmp(val, "http", 5))
          opt.ps.type = PROXY_HTTP;
        else if(!strncasecmp(val, "socks", 6)) 
          opt.ps.type = PROXY_SOCKS;
        else
          opt.ps.type = PROXY_OFF;
      }
      else if(!strncasecmp(com, "proxy_host", 11)) {
        if(get_ip_addr(val, &opt.ps.ip) == -1) {
            printout(vLESS, _("Warning: "));
            printout(vLESS, _("`%s' could not be resolved. "), val);
            printout(vLESS, _("Disabling proxy support.\n"));
            opt.ps.type = 0;
        } 
      }
      else if(!strncasecmp(com, "proxy_port", 11))
          opt.ps.port = atoi(val);
      else if(!strncasecmp(com, "proxy_user", 11))
          opt.ps.user = cpy(val);
      else if(!strncasecmp(com, "proxy_pass", 11))
          opt.ps.pass = cpy(val);
      else if(!strncasecmp(com, "proxy_bind", 11))
          opt.ps.bind = !strncasecmp(val, "on", 3);
      else if(!strncasecmp(com, "passwordfile", 13) || !strncasecmp(com, "password_file", 14))
          read_password_file(val);
      else return -1;
      return 0;
    case 'r':
        if(!strncasecmp(com, "rate", 5)) {
            return 0;
        } else if(!strncasecmp(com, "retry_count", 12))
            opt.retry = atoi(val);
        else return -1;
        return 0;
  case 's':
      if(!strncasecmp(com, "sort_urls", 10))
        opt.sorturls = !strncasecmp(val, "on", 3);
      else return -1;
      return 0;
  case 't':
      if(!strncasecmp(com, "timeout", 8))
        socket_set_default_timeout(atoi(val));
      else if(!strncasecmp(com, "timestamping", 13))
        return 0;
      else if(!strncasecmp(com, "timeoffset", 11))
        opt.time_offset     = atoi(val);
      else if(!strncasecmp(com, "timeoffset", 11))
        opt.time_deviation  = atoi(val);
      else if(!strncasecmp(com, "transfer_type", 14)) {
        return 0;
      } else return -1;
      return 0;
  case 'v':
      if(!strncasecmp(com, "verbosity", 10)) {
        char * levels[] = {"quite", "less", "normal", "more", "debug"};
        int i;
        for(i=0;i<5;i++)
          if(!strcasecmp(val, levels[i])) {
            opt.verbose = i;
            return 0;
          }
        return -2;
      } else return -1;
  case 'w':
      if(!strncasecmp(com, "wait_retry", 11))
        opt.retry_interval = atoi(val);
      else return -1;
      return 0;
  }
  return -1;
}

#define NETRC_FILE_NAME ".netrc"

/* read the netrc file and store its contents in a linked list */
void read_netrc_file(void) {
    /* Find ~/.netrc.  */
    char *path, *home;
    acc_t *netrc_list, *l;

    home = home_dir ();
    if (!home)
    return;

    path = (char *) malloc(strlen (home) + 1 +
               strlen (NETRC_FILE_NAME) + 1);
    if (!path)
    return;

    sprintf (path, "%s/%s", home, NETRC_FILE_NAME);
    free (home);

    if (!file_exists(path)) {
    printout(vMORE, _("netrc file '%s' cannot be read. skipping\n"), path);
    return;
    }

    printout(vDEBUG, "Reading netrc file '%s'", path);
    netrc_list = parse_netrc (path);
    free(path);

    /* If nothing to do...  */
    if (!netrc_list)
    return;

    for (l = netrc_list; l; l = l->next) {
    if (!l->host)
        continue;
    opt.pl = password_list_add(opt.pl, cpy(l->host), cpy(l->acc), cpy(l->passwd));
    printout(vDEBUG, "added %s:%s@%s to the password-list (%x)\n", l->acc, l->passwd, l->host, opt.pl);
    }

    free_netrc(netrc_list);
}

/* read the password file and store its contents in a linked list */
void read_password_file(char * f) {
    FILE * fp;
    char * line;
    char * tmp;
    char * user;
    char * pass;
    if(!file_exists(f)) {
        printout(vMORE, _("password_file '%s' cannot be read. skipping\n"), f);
        return;
    }
    printout(vNORMAL, _("Warning: You are using a wput password file. This is deprecated!\n"
                "         Please consider switch to the widely used netrc-files.\n"));
    printout(vDEBUG, "Reading password-file '%s'", f);
    fp = fopen(f, "r");
    while( (tmp = line = read_line(fp)) ) {
        while(isspace(*tmp) && *tmp) tmp++;
        strtok(tmp, "\t");
        user = strtok(NULL, "\t");
        pass = strtok(NULL, "\t");
        if(*tmp == 0 || *tmp == '#' || user == NULL || pass == NULL) {
            free(line);
            continue;
        }
        if(pass[strlen(pass)-2] == '\r') pass[strlen(pass)-2] = 0;
        else if(pass[strlen(pass)-1] == '\n') pass[strlen(pass)-1] = 0;
        opt.pl = password_list_add(opt.pl, cpy(tmp), cpy(user), cpy(pass));
        printout(vDEBUG, "added %s:%s@%s to the password-list (%x)\n", user, pass, tmp, opt.pl);
        free(line);
    }
}
/* reads a wputrc file. parses its options and gives error-reports if necessary */
void readwputrc(char * f) {
    FILE * fp;
    char * file;
    char * line;
    int    ln   = 1;

    if(f == NULL) {
        char * home = home_dir();
        file = malloc(strlen(home) + 10); /* home + slash + (.wputrc | wput.ini) + 0-char */
        sprintf(file, "%s/%s", home, WPUTRC_FILENAME);
        free(home);
    } else file = cpy(f);

    printout(vDEBUG, "Reading wputrc-file: %s\n", file);

    if(!file_exists(file)) {
        printout(vDEBUG, "wputrc-file '%s' is not readable. skipping.\n", file);
        free(file);
        return;
    }

    fp = fopen (file, "r");
    if(!fp) {
      printout(vLESS, _("Fatal error while opening '%s': %s\n"), file, strerror (errno));
      free(file);
      return;
    }

  while ((line = read_line (fp))) {
    char * tmp = line;
    char * com;
    char * val;
    /* skip leading spaces */
    while(isspace(*tmp) && *tmp) tmp++;

    /* discard comment lines */
    if(*tmp == '#' || *tmp == ';' || *tmp == 0) {
        free(line);
        continue;
    }
    com = tmp;

    /* search for space, end or '=' */
    while(!isspace(*tmp) && *tmp && *tmp != '=') tmp++;
    *tmp++= 0;

    while((isspace(*tmp) || *tmp == '=') && *tmp) tmp++;
    val = tmp;

    /* suppress the new-line-char */
    while(*tmp != 0 && *tmp != '\n') tmp++;
    if(*(tmp-1) == '\r') *(tmp-1) = 0;
    else                 * tmp    = 0;

    /* we mis-use tmp to store the ret_val, and print a message if something was not parse-able */
    tmp = (char *) set_option(com, val);
    if(tmp == (char *) -1) printout(vLESS, _("%s#%d: Option '%s' not recognized\n"), file, ln, com);
    if(tmp == (char *) -2) printout(vLESS, _("%s#%d: Unknow value '%s' for '%s'\n"), file, ln, val, com);
    free(line);
    ln++;
  }
  free(file);
}

/* parses all the commandline-options available. */
void commandlineoptions(int argc, char * argv[]){
    int c;

    int option_index = 0;
    static struct option long_options[] =
      {
        {"append-output", 1, 0, 'a'},    //0
        {"ascii", 0, 0, 'A'},
        {"background", 0, 0, 'b'},
        {"basename", 1, 0, 0},
        {"binary", 0, 0, 'B'},
        {"bind-address", 1, 0, 0},       //5
        {"compile-options", 0, 0, 0},    
        {"debug", 0, 0, 'd'},
        {"dont-continue", 0, 0, 0},
        {"force-tls", 0, 0, 0},
        {"help", 0, 0, 'h'},             //10
        {"input-file", 1, 0, 'i'},      
        {"input-pipe", 1, 0, 'I'},      
        {"less-verbose", 0, 0, 0},      
        {"limit-rate", 1, 0, 'l'},
        {"no-directories", 0, 0, 0},     //15
        {"output-file", 1, 0, 'o'},     
        {"port-mode", 0, 0, 'p'},       
        {"proxy", 1, 0, 'Y'},           
        {"proxy-user", 1, 0, 0},
        {"proxy-pass", 1, 0, 0},         //20
        {"quiet", 0, 0, 'q'},           
        {"random-wait", 0, 0, 0},       
        {"remove-source-files", 0, 0, 'R'},
        {"reupload", 0, 0, 'u'},        
        {"script", 1, 0, 'S'},           //25
        {"skip-existing", 0, 0, 0},     
        {"skip-larger", 0, 0, 0},       
        {"sort", 0, 0, 's'},            
        {"timeoffset", 1, 0, 0},        
        {"timeout", 1, 0, 'T'},          //30
        {"timestamping", 0, 0, 'N'},    
        {"tries", 1, 0, 't'},           
        {"use-proxy", 1, 0, 'Y'},       
        {"verbose", 0, 0, 'v'},
        {"version", 0, 0, 'V'},          //35
        {"wait", 1, 0, 'w'},            
        {"waitretry", 1, 0, 0},         
        {"chmod", 1, 0, 'm'},
        {0, 0, 0, 0}                    
      };
    while (1)
    {
        c = getopt_long (argc, argv, "Y:Vhbo:a:dqvn:i:t:T:w:ps",
                           long_options, &option_index);

        if (c == -1)
                break;

        switch (c)
        {
        case 0:
            switch(option_index) {
            case 13:  //less-verbose
                opt.verbose--;                            break;
            case  5: set_option("bind-address", optarg);  break;
            case 22: //random-wait
                opt.random_wait = 1;                      break;
            case 17: //port-mode
                opt.portmode = 1;                         break;
            case 19: set_option("proxy_user", optarg);    break;
            case 20: set_option("proxy_pass", optarg);    break;
            case  6: //compile-options
                fprintf(opt.output, "wdel version: %s\n\n#defined options:\n", version);
#ifdef TIMER_GETTIMEOFDAY 
                fprintf(opt.output, "TIMER_GETTIMEOFDAY\n");
#endif
#ifdef HAVE_SSL
                fprintf(opt.output, "HAVE_SSL\n");
#endif
                fprintf(opt.output, "\nUsing %d-Bytes for off_t\n", sizeof(off_t));
                exit(0);
            case 29: //timeoffset
                set_option("timeoffset", optarg);                   break;
#ifdef HAVE_SSL
            case 9: //force-tls
                set_option("force_tls", "on");                      break;
#endif
            case  3: //basename
                opt.basename = optarg;                              break;
            case 37: //waitretry
                opt.retry_interval = atoi(optarg);                  break;
            default:
                fprintf(stderr, _("Option %s should not appear here :|\n"), long_options[option_index].name);
            }
            break;
        case 'o': 
        case 'a': opt.output = fopen(optarg, (c == 'o') ? "w" : "a");
                  if(opt.output == NULL) { perror(_("Unable to open logfile")); exit(4); }
                  break;
#ifndef WIN32
        case 'b': opt.background = 1;           break;
#endif
        case 't': 
                  opt.retry = atoi(optarg) + 1;
                  if(opt.retry <= 0) opt.retry = -1;
                  break;
        case 'T': socket_set_default_timeout(atoi(optarg));
                                                break;
        case 'p': opt.portmode = 1;             break;
        case 'n':
                  if(optarg[0] == 'v')      opt.verbose--;
                  break;
        case 'v': opt.verbose++;                break;
        case 'q': opt.verbose=0;                break;
        case 'd': opt.verbose=vDEBUG;           break;
        case 'w': opt.wait   =atoi(optarg);     break;
        case 's': opt.sorturls = 1;             break;
        case 'i': 
                  if(!strncmp(optarg, "-", 2)) {
                    printout(vDEBUG, "Reading URLs from stdin\n");
                    opt.input = stdin;
                  } else {
                    printout(vDEBUG, "Reading URLs from `%s'\n", optarg);
                    opt.input = fopen(optarg, "r");
                    if(opt.input == NULL) {
                        perror(optarg);
                        exit(4);
                    }
                  }
                  break;
        case 'Y': set_option("proxy", optarg);  break;
        case 'V':
            fprintf(opt.output, _("wdel version: %s\n"), version);
            exit(0);
        case 'h':
        default:
            fprintf(stderr, _("Usage: wdel [options] [url [file]...]...\n"
"  url        ftp://[username[:password]@]hostname[:port][/[path/][file]]\n\n"
"Startup:\n"
"  -V, --version         Display the version of wdel and exit.\n"
"  -h, --help            Print this help-screen\n"));
#ifndef WIN32
            fprintf(stderr, _(
"  -b, --background      go to background after startup\n"));
#endif
            fprintf(stderr, "\n");

            fprintf(stderr, _(
"Logging and input file:\n"
"  -o,  --output-file=FILE      log messages to FILE\n"
"  -a,  --append-output=FILE    append log messages to FILE\n"
"  -q,  --quiet                 quiet (no output)\n"
"  -v,  --verbose               be verbose\n"
"  -d,  --debug                 debug output\n"
"  -nv, --less-verbose          be less verbose\n"
"  -i,  --input-file=FILE       read the URLs from FILE\n"
"  -s,  --sort                  sorts all input URLs by server-ip and path\n"
"       --basename=PATH         snip PATH off each file when appending to a URL\n"
"\n"));
            fprintf(stderr, _(
"Connection:\n"
"       --bind-address=ADDR     bind to ADDR (hostname or IP) on local host\n"
"  -t,  --tries=NUMBER          set retry count to NUMBER (-1 means infinite)\n"
"  -T,  --timeout=10th-SECONDS  set various timeouts to 10th-SECONDS\n"
"  -w,  --wait=10th-SECONDS     wait 10th-SECONDS between uploads. (default: 0)\n"
"       --random-wait           wait from 0...2*WAIT secs between uploads.\n"
"       --waitretry=SECONDS     wait SECONDS between retries of an upload\n"
"  -Y,  --proxy=http/socks/off  set proxy type or turn off\n"
"       --proxy-user=NAME       set the proxy-username to NAME\n"
"       --proxy-pass=PASS       set the proxy-password to PASS\n"
"\n"));
            fprintf(stderr, _(
"FTP-Options:\n"
"  -p,  --port-mode             no-passive, turn on port mode ftp (def. pasv)\n"));

#ifdef HAVE_SSL
            fprintf(stderr, _(
"       --force-tls             force the useage of TLS\n"));
#endif
            fprintf(stderr, _(
"\n"
"See wdel(1) for more detailed descriptions of the options.\n"
"Report bugs and suggestions via SourceForge at\n"
"http://sourceforge.net/tracker/?group_id=141519\n"));
            exit(0);
        }
    }

    while(optind < argc) {
        if(!strncmp(argv[optind], "ftp://", 6))
            wdel_queue_add_entry(NULL, cpy(argv[optind]));
        else
            wdel_queue_add_file(cpy(argv[optind]));
        optind++;
    }
}
